File: 040_AddTwoNumbers.cpp
// Topic: Add Two Numbers
// Description: Add two numbers represented by linked lists (digits stored in reverse order).
// Approach: Iterate lists, sum digits with carry, build result list.

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
{
    ListNode dummy(0), *tail = &dummy;
    int carry = 0;
    while (l1 || l2 || carry)
    {
        int s = carry + (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
        carry = s / 10;
        tail->next = new ListNode(s % 10);
        tail = tail->next;
        if (l1)
            l1 = l1->next;
        if (l2)
            l2 = l2->next;
    }
    return dummy.next;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // Test 1
    {
        ListNode* a = new ListNode(2);
        a->next = new ListNode(4);
        a->next->next = new ListNode(3);
        ListNode* b = new ListNode(5);
        b->next = new ListNode(6);
        b->next->next = new ListNode(4);
        ListNode* res = addTwoNumbers(a, b);
        for (ListNode* p = res; p; p = p->next)
        {
            cout << p->val;
            if (p->next)
                cout << " ";
        }
        cout << "\n";
    }
    // Test 2
    {
        ListNode* a = new ListNode(0);
        ListNode* b = new ListNode(0);
        ListNode* res = addTwoNumbers(a, b);
        for (ListNode* p = res; p; p = p->next)
        {
            cout << p->val;
            if (p->next)
                cout << " ";
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(max(n,m))
Space Complexity: O(max(n,m)) for result
*/

File: 040_AddTwoNumbers.py
# Topic: Add Two Numbers
# Description: Add two numbers represented by linked lists (digits stored in reverse order).
# Approach: Iterate lists, sum digits with carry, build result list.

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        tail = dummy
        carry = 0
        while l1 or l2 or carry:
            s = carry + (l1.val if l1 else 0) + (l2.val if l2 else 0)
            carry = s // 10
            tail.next = ListNode(s % 10)
            tail = tail.next
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        return dummy.next


def print_list(head):
    vals = []
    while head:
        vals.append(str(head.val))
        head = head.next
    print(' '.join(vals))


if __name__ == "__main__":
    # Test 1
    a = ListNode(2, ListNode(4, ListNode(3)))
    b = ListNode(5, ListNode(6, ListNode(4)))
    result = Solution().addTwoNumbers(a, b)
    print_list(result)

    # Test 2
    a = ListNode(0)
    b = ListNode(0)
    result = Solution().addTwoNumbers(a, b)
    print_list(result)

"""
Time Complexity: O(max(n,m))
Space Complexity: O(max(n,m)) for result
"""
